### DISCUSSION TITLE
ValueError: object arrays are not supported in port.optimization

### DISCUSSION THREAD

**Discussion by pfeliciano87** [dcajasn/Riskfolio-Lib#53] [Q&A]
Dear dcajasn, congratulations for coding this module, it is really helpfull and unique.
I installed the 1.0.0 version because when I install the 2.0.0 version I get an error when I "import riskfolio as rp".
The error is related to astropy: "ImportError: DLL load failed: The parrameter is incorrect."

WIth riskfolio 1.0.0 version I don't get this error as this version doesn't need astropy. 
I can "import riskfolio as rp". But then I get an error: "ValueError: object arrays are not supported".

Here is the code:
`# Import necessary libraries
import pandas as pd
import riskfolio as rp
import numpy as np

# Import data as dataframe
columns_df = ['china_eq','em_eq','euro_eq','world_small_cap','us_eq','china_bd','euro_corp','euro_gov','us_glbl_agg','global_hy','em_debt','cmdties']
df_aa = pd.read_excel('asset_classes_regimes.xlsx',index_col=0)

# Filtering the dataframe and creating new dataframe
df_aa_stress2 = df_aa[(df_aa.sectors==1) | (df_aa.forex==1)]
df_aa_normal2 = df_aa[~((df_aa.sectors==1) | (df_aa.forex==1))]

df_rets_normal = pd.DataFrame(df_aa_normal2[df_aa_normal2.columns[0:12]])
df_rets_stress = df_aa_stress2[df_aa_stress2.columns[0:12]]

#Assigning new columns' names
df_rets_normal.columns = columns_df
df_rets_stress.columns = columns_df

# Building the portfolio object
port = rp.Portfolio(returns=df_rets_normal)

model='Classic' # Could be Classic (historical), BL (Black Litterman) or FM (Factor Model)
rm = 'MV' # Risk measure used, this time will be variance
obj = 'Sharpe' # Objective function, could be MinRisk, MaxRet, Utility or Sharpe
hist = True # Use historical scenarios for risk measures that depend on scenarios
rf = 0 # Risk free rate
l = 3 # Risk aversion factor, only useful when obj is 'Utility'

w = port.optimization(model=model, rm=rm, obj=obj, rf=rf, l=l, hist=hist)
`
When I do "port.returns" I get a dataframe with the returns.

The error I get in port.optimization is in the image attached.

Thank you very much for your help here.
![image](https://user-images.githubusercontent.com/96258983/146426621-ceacba60-c196-4c09-b4f4-663455d89001.png)

**Comment by pfeliciano87** [comment]
I am using Python 3.7.7. All the dependencies are installed.

**Reply by dcajasn** [reply]
Hi @pfeliciano87 you need to install astropy to get access to the lastest versions, it is easy when you use conda or pip. Your error is because your data is a string type, not a number type. On the other hand, this is not a channel to comunicate errors, if you have an error please raise an issue, but first you must be sure that is an error of Riskfolio-Lib not an error of your data.

**Reply by pfeliciano87** [reply]
Hi @dcajasn . thank you for your quick reply. Sorry if I posted this message in the wrong place. I really don't think the problem was with my dataframe. This morning I just managed to execute the optimization with a work around. The dataframe is the same, no change. The work around was:
`
# Importing libraries
import pandas as pd
import riskfolio as rp #version 1.0.0

# Importing dataframe, filtering to have 2 scenarios, change of columns' names
df_aa = pd.read_excel('asset_classes_regimes.xlsx',index_col=0)
df_aa_stress2 = df_aa[(df_aa.sectors==1) | (df_aa.forex==1)]
df_aa_normal2 = df_aa[~((df_aa.sectors==1) | (df_aa.forex==1))]

df_rets_normal = pd.DataFrame(df_aa_normal2[df_aa_normal2.columns[0:12]])
df_rets_stress = df_aa_stress2[df_aa_stress2.columns[0:12]]

df_rets_normal.columns = columns_df
df_rets_stress.columns = columns_df


# Creation of port instance and optimization
port = rp.Portfolio(returns=df_rets_normal)

# This was the workaround
custom_cov = rp.ParamsEstimation.covar_matrix(port.returns)
port.cov = custom_cov # Custom covariance matrix.

# Estimate optimal portfolio:
model='Classic' # Could be Classic (historical), BL (Black Litterman) or FM (Factor Model)
rm = 'MV' # Risk measure used, this time will be variance
obj = 'Sharpe' # Objective function, could be MinRisk, MaxRet, Utility or Sharpe
hist = True # Use historical scenarios for risk measures that depend on scenarios
rf = 0 # Risk free rate
l = 3 # Risk aversion factor, only useful when obj is 'Utility'

w = port.optimization(model=model, rm=rm, obj=obj, rf=rf, l=l, hist=hist)

# IT WORKED!`

**Reply by asicoderOfficial** [reply]
Hi!
I have the same problem. Followed Tutorial_1.ipynb, but I can not make it work, even though I have no strings in my DataFrame and everything seems to be as in the tutorial.
I tried your workaround @pfeliciano87 but it does not work, it raises the error:

```
Traceback (most recent call last):
  File "/home/asicoder/gitrepos/TradingBot/src/portfolio_creation_basic.py", line 123, in <module>
    a()
  File "/home/asicoder/gitrepos/TradingBot/src/portfolio_creation_basic.py", line 59, in a
    res = port.optimization()
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/riskfolio/Portfolio.py", line 1154, in optimization
    devconstraints = [cv.SOC(g, G.T @ w)]
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/expressions/expression.py", line 717, in __array_ufunc__
    return ufunc_handler(self, args[0])
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/expressions/expression.py", line 74, in <lambda>
    np.matmul: lambda self, a: self.__rmatmul__(a),
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/expressions/expression.py", line 49, in cast_op
    return binary_op(self, other)
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/expressions/expression.py", line 647, in __rmatmul__
    return cvxtypes.matmul_expr()(other, self)
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/atoms/affine/binary_operators.py", line 50, in __init__
    super(BinaryOperator, self).__init__(lh_exp, rh_exp)
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/atoms/atom.py", line 51, in __init__
    self._shape = self.shape_from_args()
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/atoms/affine/binary_operators.py", line 120, in shape_from_args
    return u.shape.mul_shapes(self.args[0].shape, self.args[1].shape)
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/utilities/shape.py", line 142, in mul_shapes
    lh_shape, rh_shape, shape = mul_shapes_promote(lh_shape, rh_shape)
  File "/home/asicoder/anaconda3/envs/main_python_env/lib/python3.9/site-packages/cvxpy/utilities/shape.py", line 108, in mul_shapes_promote
    raise ValueError("Incompatible dimensions %s %s" % (
ValueError: Incompatible dimensions (148, 148) (1, 1)
```

I can not see what is going on.

**Reply by dcajasn** [reply]
Hi @asicoderOfficial

It seems like a problem of portfolio_creation_basic.py, not a riskfolio-lib problem. 

Best,
Dany

**Reply by wytxty** [reply]
Hi, I think u may miss the line like:    
    method_mu = 'hist'
    method_cov = 'hist'

    port.assets_stats(method_mu=method_mu, method_cov=method_cov, d=0.94)
